packages/plugin-react/src/fast-refresh.js

```diff
@@ -0,0 +1,109 @@
+import fs from 'fs'
+
+export const runtimePublicPath = '/@react-refresh'
+
+const runtimeFilePath = require.resolve(
+  'react-refresh/cjs/react-refresh-runtime.development.js'
+)
+
+export const runtimeCode = `
+const exports = {}
+${fs.readFileSync(runtimeFilePath, 'utf-8')}
+function debounce(fn, delay) {
+  let handle
+  return () => {
+    clearTimeout(handle)
+    handle = setTimeout(fn, delay)
+  }
+}
+exports.performReactRefresh = debounce(exports.performReactRefresh, 16)
+export default exports
+`
+
+export const preambleCode = `
+import RefreshRuntime from "__BASE__${runtimePublicPath.slice(1)}"
+RefreshRuntime.injectIntoGlobalHook(window)
+window.$RefreshReg$ = () => {}
+window.$RefreshSig$ = () => (type) => type
+window.__vite_plugin_react_preamble_installed__ = true
+`
+
+const header = `
+import RefreshRuntime from "${runtimePublicPath}";
+
+let prevRefreshReg;
+let prevRefreshSig;
+
+if (!window.__vite_plugin_react_preamble_installed__) {
+  throw new Error(
+    "@vitejs/plugin-react-refresh can't detect preamble. Something is wrong. " +
+    "See https://github.com/vitejs/vite-plugin-react/pull/11#discussion_r430879201"
+  );
+}
+
+if (import.meta.hot) {
+  prevRefreshReg = window.$RefreshReg$;
+  prevRefreshSig = window.$RefreshSig$;
+  window.$RefreshReg$ = (type, id) => {
+    RefreshRuntime.register(type, ${JSON.stringify(id)} + " " + id)
+  };
+  window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
+}`.replace(/[\n]+/gm, '')
+
+export function addRefreshWrapper(code, accept) {
```

Comment:
**suggestion:** We may need to disable `@typescript-eslint/explicit-module-boundary-types` for `js` files :thinking:

Reply:
Done via https://github.com/vitejs/vite/pull/3589

Please rebase after PR merged